### PR TITLE
CPM-228: Add skipped product with no diff for rule engine execution summary

### PR DESCRIPTION
**Pull request by cemorin**
**Description (for Contributor and Core Developer)**
Add number of skipped products on product rule execution summary

Screenshot from 2022-02-21 11-39-57

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php [hunk 1]
```diff
@@ -85,6 +96,8 @@ function (ProductInterface $product): bool {
             );
         }
 
+        $this->eventDispatcher->dispatch(new SkippedProductSaveAction($skippedProducts));
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php:99]:
```suggestion
        if ([] !== $skippedProducts) {
            $this->eventDispatcher->dispatch(new SkippedProductSaveAction($skippedProducts));
        }
```

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php [hunk 2]
```diff
@@ -115,6 +128,7 @@ function (ProductInterface $product): bool {
                 StorageEvents::POST_SAVE
             );
         }
+//        $this->eventDispatcher->dispatch(new SkippedProductSaveAction($skippedProducts));
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php:131]: to remove

File: src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php:6]: Not the good header

File: src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Pim\Enrichment\Bundle\Event;
+
+use Akeneo\Tool\Component\Batch\Event\EventInterface;
+
+class SkippedProductSaveAction implements EventInterface
+{
+    protected array $subjectSet;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php:20]:
```suggestion
    /** @var ProductInterface[] */
    private array $products;
```

File: src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Pim\Enrichment\Bundle\Event;
+
+use Akeneo\Tool\Component\Batch\Event\EventInterface;
+
+class SkippedProductSaveAction implements EventInterface
```
[the remark this answers is not in the document]
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php:18]: There's already a similar event for skipped products due to errors called `SkippedActionForSubjectEvent`.
Would `SkippedActionNoDiffForSubjectEvent` be better?

[the remark this answers is not in the document]
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php]: ProductsWereSkippedDuringSave

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Bundle/Event/SkippedProductSaveAction.php]: I added ProductModelsWereSkippedDuringSave is that ok?

File: src/Akeneo/Pim/Enrichment/Bundle/Event/ProductModelsWereSkippedDuringSave.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class ProductModelsWereSkippedDuringSave implements EventInterface
+{
+    /** @var ProductModelInterface[] */
+    private array $subjectSet;
+
+    public function __construct(array $products)
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Event/ProductModelsWereSkippedDuringSave.php:19]:
```suggestion
    public function __construct(array $productModels)
```

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Event/ProductModelsWereSkippedDuringSave.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+use Akeneo\Tool\Component\Batch\Event\EventInterface;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class ProductModelsWereSkippedDuringSave implements EventInterface
+{
+    /** @var ProductModelInterface[] */
+    private array $subjectSet;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Event/ProductModelsWereSkippedDuringSave.php:17]: maybe rather skippedProductModels or skippedEntities?

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelSaver.php [hunk 8]
```diff
@@ -63,7 +65,16 @@ public function saveAll(array $productModels, array $options = [])
...
+        $skippedProductModels = [];
         if (true !== ($options['force_save'] ?? false)) {
+            $skippedProductModels = array_values(
+                array_filter(
+                    $productModels,
+                    function (ProductModelInterface $productModel): bool {
+                        return !$productModel->isDirty();
+                    }
+                )
+            );
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelSaver.php:77]: I'd rather have `$productModels` and `$skippedProductModels` calculated during the same loop, e.g:
```php
$productModelsToSave = $skippedProductModels = [];
foreach ($productModels as $productModel) {
    if ($productModel->isDirty()) {
        $productModelsToSave[] = $productModel;
    } else {
        $skippedProductModels[] = $productModel;
    }
}
```

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php [hunk 9]
```diff
@@ -74,7 +76,16 @@ public function saveAll(array $products, array $options = [])
...
+        $skippedProducts = [];
         if (true !== ($options['force_save'] ?? false)) {
+            $skippedProducts = array_values(
+                array_filter(
+                    $products,
+                    function (ProductInterface $product): bool {
+                        return !$product->isDirty();
+                    }
+                )
+            );
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php:88]: same, you can calculate skipped and not skipped in a single loop

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php [hunk 10]
```diff
@@ -85,6 +96,10 @@ function (ProductInterface $product): bool {
             );
         }
 
+        if ([] !== $skippedProducts) {
+            $this->eventDispatcher->dispatch(new ProductsWereSkippedDuringSave($skippedProducts));
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductSaver.php:100]: please update the Product and ProductModelSaver specs accordingly

**Pull request closed without merging**